### DISCUSSION TITLE
Question on Distance2PoincareHyperplanes

### DISCUSSION THREAD

**Discussion by molbap** [geoopt/geoopt#176] [Q&A]
In the example notebook hyperbolic_multiclass_classification, the layer Distance2PoincareHyperplanes defines the distance in the forward pass as 

```
distance = self.ball.dist2plane(
            x=input_p, p=points, a=points, signed=self.signed, dim=-self.n - 2
        )
```

Where points are defined as

```

        self.points = geoopt.ManifoldParameter(
            torch.empty(num_planes, plane_shape), manifold=self.ball
        )
        points = self.points.permute(1, 0)
        points = points.view(points.shape + (1,) * self.n)

```

My question is, **why are the parameters a and p equal?** In the dist2plane definition, a belongs to the tangent space of the manifold, and p is a point on the manifold itself. It is not obvious to me why these two entries should be equal.

**Comment by ferrine** [comment]
The idea was to remove overparametrization. You can get any plane with a single parameter

**Reply by jin-zhe** [reply]
Hi may I know how and why that re-parameterization works? Based on my understanding, we need minimally those two points to mathematically define a plane with `a` being orthogonal to the plane and `p` being a point on the plane.

Unless here we are saying we want the orthogonal to also be the point on the plane?

**Reply by ferrine** [reply]
For every single hyperplane you may find the closest point there to the origin. Log0 of the point is collinear to the tangent vector that is orthogonal to the hyperplane at this point. But I put a concern there because overparametrization is not always bad and you may use both alternatives to figure out which is better

**Reply by ferrine** [reply]
And yes, orthogonal vector `a` is ALWAYS a tangent vector pointing from the point `p`. To hold all the intuition of the definition you should use p and transp(a0, 0 -> p), meaning "moving `p` requires `a` to be also moved"

**Comment by jin-zhe** [comment]
> In the example notebook hyperbolic_multiclass_classification

Hi may I know where is this notebook? I've tried looking for it under the docs but couldn't find it.

**Reply by MichaelGoodale** [reply]
https://github.com/geoopt/geoopt/blob/master/examples/hyperbolic_multiclass_classification.ipynb Here you go!

**Comment by jin-zhe** [comment]
Hi @ferrine does the `dist2plane` method in geoopt accord to "MLR in the Poincaré ball" formulation derived by Ganea et. al. in the paper "Hyperbolic Neural Networks"?

**Reply by ferrine** [reply]
Yes, it does and extended to support positive curvature. It also can be scaled by the tangent norm (originally, the scale of tangent does not make sense).
A more detailed doc is [here](https://geoopt.readthedocs.io/en/latest/extended/stereographic.html#geoopt.manifolds.stereographic.math.dist2plane)

**Reply by jin-zhe** [reply]
One more point of discussion if I may.

In the original TF implementation by Ganea, [gradients for plane parameters are clipped at norms of 1](https://github.com/dalab/hyperbolic_nn/blob/master/hyp_rnn.py#L576) so as to facilitate rescaling them into Riemannian gradients. Do we have to do something similar when using geoopt's `dist2plane` with `RiemannianSGD` ? 

Also are parameters ensured to be kept within the unit ball after updates using `RiemannianSGD`?

**Reply by jin-zhe** [reply]
Bump^

**Reply by ferrine** [reply]
They are projected to the unit ball by implentation (you can turn it off). You can clip gradients manually, with a hook as you do it with a regular pytorch tensor. However, the concern would be the unit norm. Is it unit in a Riemannian sense or in a simpler numeric sense.
